storage-proofs/post/src/fallback/vanilla.rs

```diff
@@ -332,9 +342,33 @@ impl<'a, Tree: 'a + MerkleTreeTrait> ProofScheme<'a> for FallbackPoSt<'a, Tree>
                             challenge_index,
                         )?;
 
-                        tree.gen_cached_proof(challenged_leaf_start as usize, Some(rows_to_discard))
+                        let proof = tree.gen_cached_proof(
+                            challenged_leaf_start as usize,
+                            Some(rows_to_discard),
+                        );
+                        let valid = if let Ok(proof) = &proof {
+                            proof.validate(challenged_leaf_start as usize)
+                                && proof.root() == priv_sector.comm_r_last
+                        } else {
+                            false
+                        };
+                        if valid {
+                            Ok(ProofOrFault::Proof(proof))
+                        } else {
+                            Ok(ProofOrFault::Fault(sector_id))
+                        }
                     })
-                    .collect::<Result<Vec<_>>>()?;
+                    .collect::<Result<Vec<_>>>()?
```

Comment:
I don't think you need to `collect()` when you loop over the iterator and are not interested in having a vector in the end.

Reply:
That's what I thought too, but sadly this was the only one of the obvious alternatives I could get to compile.